src/Orleans.Runtime/GrainDirectory/GrainLocator.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Orleans.Configuration;
+using Orleans.GrainDirectory;
+
+namespace Orleans.Runtime.GrainDirectory
+{
+    /// <summary>
+    /// Implementation of <see cref="IGrainLocator"/> that uses an <see cref="IGrainDirectory"/> store.
+    /// </summary>
+    internal class GrainLocator : IGrainLocator
+    {
+        private readonly IGrainDirectory grainDirectory;
+        private readonly InClusterGrainLocator inClusterGrainLocator;
+        private readonly IGrainDirectoryCache cache;
+
+        public GrainLocator(IGrainDirectory grainDirectory, InClusterGrainLocator inClusterGrainLocator)
+        {
+            this.grainDirectory = grainDirectory;
+            this.inClusterGrainLocator = inClusterGrainLocator;
+            this.cache = new LRUBasedGrainDirectoryCache(GrainDirectoryOptions.DEFAULT_CACHE_SIZE, GrainDirectoryOptions.DEFAULT_MAXIMUM_CACHE_TTL);
+        }
+
+        public async Task<List<ActivationAddress>> Lookup(GrainId grainId)
+        {
+            if (grainId.IsClient)
+                return await this.inClusterGrainLocator.Lookup(grainId);
+
+            var results = new List<ActivationAddress>();
+
+            var entry = await this.grainDirectory.Lookup(grainId.ToParsableString());
+
+            if (entry == null)
+                return results;
+
+            results.Add(ConvertToActivationAddress(entry));
+            this.cache.AddOrUpdate(grainId, results.Select(item => Tuple.Create(item.Silo, item.Activation)).ToList(), 0);
+
+            return results;
+        }
+
+        public async Task<ActivationAddress> Register(ActivationAddress address)
+        {
+            if (address.Grain.IsClient)
+                return await this.inClusterGrainLocator.Register(address);
+
+            var result = await this.grainDirectory.Register(ConvertToGrainAddress(address));
+            var activationAddress = ConvertToActivationAddress(result);
+            this.cache.AddOrUpdate(
+                activationAddress.Grain,
+                new List<Tuple<SiloAddress,ActivationId>>() { Tuple.Create(activationAddress.Silo, activationAddress.Activation) },
```

Comment:
nit: in a future PR, we should replace this Tuple with either a ValueTuple or a purpose-built struct

Reply:
Benefit of a ValueTuple is automatic GetHashCode/Equality, etc, if that matters.